File: server/app/schemas/inbox.py
"""Schémas de sortie du centre de notifications in-app (inbox)."""
from marshmallow import EXCLUDE, Schema, fields, validate


class NotificationSchema(Schema):
    class Meta:
        unknown = EXCLUDE
    id = fields.Int(dump_only=True)
    type = fields.Str()
    title = fields.Str()
    body = fields.Str(allow_none=True)
    is_read = fields.Bool()
    meta = fields.Dict()
    created_at = fields.Str()


class NotificationListSchema(Schema):
    class Meta:
        unknown = EXCLUDE
    notifications = fields.List(fields.Nested(NotificationSchema))


class UnreadCountSchema(Schema):
    class Meta:
        unknown = EXCLUDE
    count = fields.Int()


class LiveAlertSchema(Schema):
    class Meta:
        unknown = EXCLUDE
    key = fields.Str()
    title = fields.Str()
    body = fields.Str()
    severity = fields.Str(description="info | warning | error")
    site_ids = fields.List(fields.Int())
    site_names = fields.List(fields.Str())


class LiveAlertListSchema(Schema):
    class Meta:
        unknown = EXCLUDE
    alerts = fields.List(fields.Nested(LiveAlertSchema))


class InboxPreferencesSchema(Schema):
    class Meta:
        unknown = EXCLUDE
    notify_menu_unpublished = fields.Bool()
    notify_menu_during_service = fields.Bool()
    notify_menu_tomorrow = fields.Bool()
    notify_traffic_drop = fields.Bool()
    notify_low_satisfaction = fields.Bool()
    notify_vote_anomaly = fields.Bool()
    notify_site_inactive = fields.Bool()
    notify_holiday_approaching = fields.Bool()
    holiday_alert_days_before = fields.Int()
    weekly_digest = fields.Bool()
    digest_day = fields.Int(validate=validate.Range(min=0, max=6))
    digest_hour = fields.Int(validate=validate.Range(min=6, max=21))

File: client/src/lib/api/inbox.ts
import { api } from './client';

export interface InboxNotification {
  id: number;
  type: 'business_alert' | 'user_action' | string;
  title: string;
  body?: string | null;
  is_read: boolean;
  meta?: Record<string, unknown> | null;
  created_at: string;
  user_id?: number | null;
}

export interface NotifPreferences {
  notify_menu_unpublished: boolean;
  notify_menu_during_service: boolean;
  notify_menu_tomorrow: boolean;
  notify_traffic_drop: boolean;
  notify_low_satisfaction: boolean;
  notify_vote_anomaly: boolean;
  notify_site_inactive: boolean;
  notify_holiday_approaching: boolean;
  holiday_alert_days_before: number;
  weekly_digest: boolean;
  /** Day of week (Monday = 0) and whole hour, in Paris, of the weekly email. */
  digest_day: number;
  digest_hour: number;
}

export interface LiveAlert {
  key: string;
  title: string;
  body: string;
  /** Sites the alert caught; empty when the rule is not site-scoped. */
  site_ids: number[];
  site_names: string[];
  severity: 'error' | 'warning' | 'info';
}

export const inboxApi = {
  list: async (): Promise<InboxNotification[]> => {
    const response = await api.get('/inbox');
    return response.data.notifications as InboxNotification[];
  },

  unreadCount: async (): Promise<number> => {
    const response = await api.get('/inbox/unread-count');
    return response.data.count as number;
  },

  markRead: async (id: number): Promise<void> => {
    await api.put(`/inbox/${id}/read`);
  },

  markAllRead: async (): Promise<void> => {
    await api.put('/inbox/read-all');
  },

  delete: async (id: number): Promise<void> => {
    await api.delete(`/inbox/${id}`);
  },

  getNotifPreferences: async (): Promise<NotifPreferences> => {
    const response = await api.get('/inbox/notification-preferences');
    return response.data as NotifPreferences;
  },

  updateNotifPreferences: async (prefs: Partial<NotifPreferences>): Promise<NotifPreferences> => {
    const response = await api.put('/inbox/notification-preferences', prefs);
    return response.data as NotifPreferences;
  },

  getLiveAlerts: async (): Promise<LiveAlert[]> => {
    const response = await api.get('/inbox/live-alerts');
    return response.data.alerts as LiveAlert[];
  },
};
